PowerEditor/src/NppBigSwitch.cpp

```diff
@@ -2089,22 +2103,147 @@ LRESULT Notepad_plus::process(HWND hwnd, UINT message, WPARAM wParam, LPARAM lPa
+				// otherwise the N++ most probably crashes itself without any tidy-up
+
+				string strLog = "Main N++ wnd is disabled by (an active modal-dlg?):  ";
+				char szBuf[MAX_PATH + 128] = { 0 };
+
+				HWND hActiveWnd = ::GetActiveWindow();
+				if (hActiveWnd)
+				{
+					if (::GetWindowTextA(hActiveWnd, szBuf, _countof(szBuf)))
+						strLog += szBuf;
+					::SendMessage(hActiveWnd, WM_CLOSE, 0, 0);
+				}
 				else
-					writeLog(nppIssueLog.c_str(), "WM_CLOSE (isQueryEndSessionStarted == true)");
+				{
+					// no active child window, so it is most probably the system dialog box class #32770 (used e.g. by the MessageBox WINAPI)
+					// - so our main hwnd here is not the PARENT but OWNER of that top-level window
+					hActiveWnd = ::GetLastActivePopup(hwnd);
+					if (hActiveWnd)
+					{
+						if (::GetWindowTextA(hActiveWnd, szBuf, _countof(szBuf)))
+							strLog += szBuf;
+						::GetClassNameA(hActiveWnd, szBuf, _countof(szBuf));
+						if (0 == lstrcmpiA("#32770", szBuf))
+							strLog += " (MessageBox)";
+						// we cannot use here the usual sending of the WM_CLOSE as it will not always work (e.g. for a MB_YESNO MessageBox)
+						if (!::EndDialog(hActiveWnd, 0))
+						{
+							strLog += "  -> EndDialog failed with ErrorCode: ";
+							strLog += std::to_string(::GetLastError());
+							// last attempt
+							::SendMessage(hActiveWnd, WM_SYSCOMMAND, SC_CLOSE, 0);
+						}
+					}
+					else
+					{
+						strLog += "???";
+					}
+				}
+
+				// re-test
+				if (::IsWindowEnabled(hwnd))
+					strLog += "  -> Main N++ wnd has been successfully reenabled.";
+
+				if (nppParam.doNppLogNulContentCorruptionIssue())
+				{
+					generic_string issueFn = nppLogNulContentCorruptionIssue;
+					issueFn += TEXT(".log");
+					generic_string nppIssueLog = nppParam.getUserPath();
+					pathAppend(nppIssueLog, issueFn);
+					writeLog(nppIssueLog.c_str(), strLog.c_str());
+				}
 			}
 
+			// TODO: here is the last opportunity to call the following WINAPI in a possible future version of the N++
+			// 
+			// flags RESTART_NO_PATCH and RESTART_NO_REBOOT are not set, so we should be restarted if terminated by an update or restart
+			//::RegisterApplicationRestart(restartCommandLine.c_str(), RESTART_NO_CRASH | RESTART_NO_HANG);
+
+			return TRUE; // nowadays, with the monstrous Win10+ Windows Update behind, is futile to try to interrupt the shutdown by returning FALSE here
+						 // (if one really needs so, there is the ShutdownBlockReasonCreate WINAPI for the rescue ...)
+		}
+
+		case WM_ENDSESSION:
+		{
+			// this message informs our app whether the session is really ending
+
+			if (nppParam.doNppLogNulContentCorruptionIssue())
+			{
+				generic_string issueFn = nppLogNulContentCorruptionIssue;
+				issueFn += TEXT(".log");
+				generic_string nppIssueLog = nppParam.getUserPath();
+				pathAppend(nppIssueLog, issueFn);
+
+				string wmesType = std::to_string(lParam);
+				if (0 == lParam)

```

Comment:
use `isShuttingDown`instead of  `0 == lParam`;

Reply:
The `lParam` here is a bit mask for different shutdown types and combinations of these. Even the `else`  here means that "isShuttingDown", I mean - the following then will look weird:
```
if (isShuttingDown)
{
    // usual shutdown
}
else
{
    // all other possible shutdown types & combinations
}
```